### PR TITLE
[Yul] Full suite tests.

### PR DESCRIPTION
**Pull request by chriseth**
This can already be merged, but of course we have to revisit it, add new stages and continuously check whether the results are satisfactory. This will also need more complicated test cases later.

### CODE REVIEW

File: test/libyul/yulOptimizerTests/fullSuite/medium.yul [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    let x := allocate(0x40)
+    mstore(array_index_access(x, 3), 2)
+}
+// ----
+// fullSuite
+// {
+//     {
+//         let _1 := 0x20
+//         let allocate_p
+//         allocate_p := mload(_3)
```
Comment (leonardoalt) [inline test/libyul/yulOptimizerTests/fullSuite/medium.yul]: Is the order wrong here? `_3` is only assigned later

Comment (chriseth) [reply test/libyul/yulOptimizerTests/fullSuite/medium.yul]: Yes, I think this is caused by the bug in the CSE. Will rebase.

Comment (chriseth) [reply test/libyul/yulOptimizerTests/fullSuite/medium.yul]: Actually yeah, this test is what caused me to investigate :)

Comment (chriseth) [reply test/libyul/yulOptimizerTests/fullSuite/medium.yul]: Updated.

File: test/libyul/yulOptimizerTests/fullSuite/medium.yul [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    pop(allocate(0x20))
+    let x := allocate(0x40)
+    mstore(array_index_access(x, 3), 2)
+}
+// ----
+// fullSuite
+// {
+//     {
+//         let _12 := 0x20
+//         let allocate__7 := 0x40
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/fullSuite/medium.yul:18]: I hope @christianparpart 's PR will change this back to `allocate_p`.

File: test/libyul/yulOptimizerTests/fullSuite/medium.yul [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+//         let allocate_p_2 := mload(allocate__7)
+//         mstore(allocate__7, add(allocate_p_2, _12))
+//         pop(allocate_p_2)
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/fullSuite/medium.yul:21]: We still need 'remove side-effect free statements'

File: test/libyul/yulOptimizerTests/fullSuite/medium.yul [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+//         let allocate_p_2_1 := mload(allocate__7)
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/fullSuite/medium.yul:22]: And of course a step that removes redundant loads and stores.

Comment (chriseth) [reply test/libyul/yulOptimizerTests/fullSuite/medium.yul:22]: But I think this will not be required for the ABI encoder.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**